src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -210,34 +201,78 @@ internal bool NullableSemanticAnalysisEnabled
         /// </summary>
         internal bool IsPeVerifyCompatEnabled => LanguageVersion < LanguageVersion.CSharp7_2 || Feature("peverify-compat") != null;
 
-        internal bool ShouldRunNullableWalker
+        /// <summary>
+        /// Returns true if nullable analysis is enabled in this compilation.
+        /// </summary>
+        /// <remarks>
+        /// Returns false if analysis is explicitly disabled for all methods; otherwise
+        /// returns true if the nullable context is enabled in the compilation options
+        /// or if there are any nullable-enabled contexts in the compilation.
+        /// </remarks>
+        internal bool IsNullableAnalysisEnabled
         {
             get
             {
-                if (!_lazyShouldRunNullableWalker.HasValue())
+                if (!_lazyShouldRunNullableAnalysis.HasValue())
                 {
-                    if (Options.NullableContextOptions != NullableContextOptions.Disable)
+                    _lazyShouldRunNullableAnalysis = isEnabled().ToThreeState();
+                }
+                return _lazyShouldRunNullableAnalysis.Value();
+
+                bool isEnabled()
+                {
+                    if (GetNullableAnalysisValue() == false)
                     {
-                        _lazyShouldRunNullableWalker = ThreeState.True;
-                        return true;
+                        return false;
                     }
-
-                    foreach (var syntaxTree in SyntaxTrees)
+                    if (Options.NullableContextOptions != NullableContextOptions.Disable)
                     {
-                        if (((CSharpSyntaxTree)syntaxTree).HasNullableEnables())
-                        {
-                            _lazyShouldRunNullableWalker = ThreeState.True;
-                            return true;
-                        }
+                        return true;
                     }
-
-                    _lazyShouldRunNullableWalker = ThreeState.False;
+                    return SyntaxTrees.Any(tree => ((CSharpSyntaxTree)tree).HasNullableEnables());
                 }
+            }
+        }
 
-                return _lazyShouldRunNullableWalker.Value();
+        /// <summary>
+        /// Returns true if nullable analysis is enabled for all methods regardless
+        /// of the actual nullable context.
+        /// If this property returns true but IsNullableAnalysisEnabled returns false,
+        /// any nullable analysis should be enabled but results should be ignored.
```

Comment:
Did you consider having a named property that reflected this logic like: `IsNullableWalkerVerificationEnabled`?

Reply:
I did consider variants of these two properties but wasn't able to find variants that simplified callers. I'm happy to revisit this later.

---
In reply to: [542548118](https://github.com/dotnet/roslyn/pull/49876#discussion_r542548118) [](ancestors = 542548118)